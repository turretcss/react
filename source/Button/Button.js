import React from 'react'
import classnames from 'classnames'

import { indicatorMap, sizeAliases } from '../constants'

const Button = ({
  Tag = 'button',
  type = 'button',
  className,
  size,
  border,
  square,
  color,
  pill,
  spinner,
  block,
  children,
  text,
  ...props
}) => (
  <Tag
    {...props}
    type={type}
    className={
      classnames(className, {
        ['button-' + (sizeAliases[size] || size)]: size,
        ['button-' + color]: color && !indicatorMap[color],
        ['button-block']: block,
        ['button-border']: border,
        ['button-pill']: pill,
        ['button-square']: square,
        ['spinner']: spinner,
        [color]: indicatorMap[color],
      }) || null
    }
    children={children || text}
  />
)

export const ButtonLink = props => <Button {...props} Tag="a" className={['button', props.className]} type={null} />

export default Button
