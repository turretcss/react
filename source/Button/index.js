export { default } from './Button'
export { default as Button, ButtonLink } from './Button'
