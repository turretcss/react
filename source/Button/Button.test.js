import React from 'react'
import renderer from 'react-test-renderer'
import { renderToStaticMarkup } from 'react-dom/server'

import Button, { ButtonLink } from './Button'

describe('Button', () => {
  it('renders', () => {
    const result = renderToStaticMarkup(<Button>Hello</Button>)
    const expected = '<button type="button">Hello</button>'
    expect(result).toEqual(expected)
  })

  it('renders with text prop', () => {
    const result = renderToStaticMarkup(<Button text="Hello" />)
    const expected = '<button type="button">Hello</button>'
    expect(result).toEqual(expected)
  })

  it('prefers children over text prop', () => {
    const result = renderToStaticMarkup(<Button text="Ignored">Hello</Button>)
    const expected = '<button type="button">Hello</button>'
    expect(result).toEqual(expected)
  })

  it('renders type prop', () => {
    const result = renderToStaticMarkup(<Button type="submit" text="Submit" />)
    const expected = '<button type="submit">Submit</button>'
    expect(result).toEqual(expected)
  })

  it('renders size prop', () => {
    const result = renderToStaticMarkup(<Button size="l" />)
    const expected = '<button type="button" class="button-l"></button>'
    expect(result).toEqual(expected)
  })

  it('renders size prop (alias)', () => {
    const result = renderToStaticMarkup(<Button size="large" />)
    const expected = '<button type="button" class="button-l"></button>'
    expect(result).toEqual(expected)
  })

  it('renders color prop (palette)', () => {
    const result = renderToStaticMarkup(<Button color="secondary" />)
    const expected = '<button type="button" class="button-secondary"></button>'
    expect(result).toEqual(expected)
  })

  it('renders color prop (indicator)', () => {
    const result = renderToStaticMarkup(<Button color="error" />)
    const expected = '<button type="button" class="error"></button>'
    expect(result).toEqual(expected)
  })

  it('renders block prop', () => {
    const result = renderToStaticMarkup(<Button block />)
    const expected = '<button type="button" class="button-block"></button>'
    expect(result).toEqual(expected)
  })

  it('renders border prop', () => {
    const result = renderToStaticMarkup(<Button border />)
    const expected = '<button type="button" class="button-border"></button>'
    expect(result).toEqual(expected)
  })

  it('renders pill prop', () => {
    const result = renderToStaticMarkup(<Button pill />)
    const expected = '<button type="button" class="button-pill"></button>'
    expect(result).toEqual(expected)
  })

  it('renders square prop', () => {
    const result = renderToStaticMarkup(<Button square />)
    const expected = '<button type="button" class="button-square"></button>'
    expect(result).toEqual(expected)
  })

  it('renders spinner prop', () => {
    const result = renderToStaticMarkup(<Button spinner />)
    const expected = '<button type="button" class="spinner"></button>'
    expect(result).toEqual(expected)
  })

  it('renders className prop', () => {
    const result = renderToStaticMarkup(<Button className="custom-do-dah" />)
    const expected = '<button type="button" class="custom-do-dah"></button>'
    expect(result).toEqual(expected)
  })

  it('spreads additional props', () => {
    const result = renderToStaticMarkup(<Button data-name="Adam" style={{ margin: 'auto' }} />)
    const expected = '<button data-name="Adam" style="margin:auto" type="button"></button>'
    expect(result).toEqual(expected)
  })

  it('renders complex button', () => {
    const tree = renderer
      .create(
        <Button
          type="submit"
          className="custom-do-dah"
          aria-role="button"
          color="primary"
          size="xs"
          block
          border
          onClick={event => {}}>
          Hello
        </Button>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('ButtonLink', () => {
  it('renders anchor tag', () => {
    const result = renderToStaticMarkup(<ButtonLink href="http://turretcss.com">Hello</ButtonLink>)
    const expected = '<a href="http://turretcss.com" class="button">Hello</a>'
    expect(result).toEqual(expected)
  })

  it('renders complex anchor', () => {
    const tree = renderer
      .create(
        <ButtonLink
          href="http://turretcss.com"
          target="_blank"
          className="custom-do-dah"
          aria-role="button"
          color="primary"
          size="xs"
          block
          border
          onClick={() => {}}>
          Hello
        </ButtonLink>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
