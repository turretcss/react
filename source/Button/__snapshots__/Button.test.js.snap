// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders complex button 1`] = `
<button
  aria-role="button"
  className="custom-do-dah button-xs button-primary button-block button-border"
  onClick={[Function]}
  type="submit"
>
  Hello
</button>
`;

exports[`ButtonLink renders complex anchor 1`] = `
<a
  aria-role="button"
  className="button custom-do-dah button-xs button-primary button-block button-border"
  href="http://turretcss.com"
  onClick={[Function]}
  target="_blank"
  type={null}
>
  Hello
</a>
`;
